File: LeetCode/Note_0232_Implement_Queue_Using_Stacks/Note_0232_Implement_Queue_Using_Stacks.py
class MyQueue(object):

    def __init__(self):
        self.array1 = []
        self.array2 = []
        self.front = None

    def push(self, x: int) -> None:
        if not self.array1:
            self.front = x
        self.array1.append(x)
        return None

    def pop(self) -> int:
        if not self.array2:
            while self.array1:
                self.array2.append(self.array1.pop())
        pop_value = self.array2[-1]
        self.array2.pop()
        return pop_value
        

    def peek(self) -> int:
        if self.array2:
            return self.array2[-1]
        return self.front

    def empty(self) -> bool:
        return not self.array1 and not self.array2
        


    # def __init__(self):
    #     self.array = []

    # def push(self, x: int) -> None:
    #     self.array.append(x)
    #     return None

    # def pop(self) -> int:
    #     if not self.array:
    #         return None
    #     pop_value = self.array[0]
    #     del self.array[0]
    #     return pop_value

    # def peek(self) -> int:
    #     return self.array[0]

    # def empty(self) -> bool:
    #     if not self.array:
    #         return True
    #     else:
    #         return False
        
   

if __name__ == '__main__':
    print("template")
    myQueue = MyQueue()
    myQueue.push(1)
    myQueue.push(2)
    print(myQueue.peek())
    print(myQueue.pop())
    print(myQueue.empty())


# T:O(n)
# S:O(n)

File: LeetCode/Note_0232_Implement_Queue_Using_Stacks/Note_0232_Implement_Queue_Using_Stacks.java
import java.util.*;

class MyQueue {

    private int front;
    private Stack<Integer> s1 = new Stack<>();
    private Stack<Integer> s2 = new Stack<>();

    public MyQueue() {
        
    }
    
    public void push(int x) {
        if (this.s1.isEmpty()) {
            this.front = x;
        }
        this.s1.push(x);
    }
    
    public int pop() {
        if (this.s2.isEmpty()) {
            while (!this.s1.isEmpty()) {
                this.s2.push(this.s1.pop());
            }
        }
        return this.s2.pop();
    }
    
    public int peek() {
        if (!this.s2.isEmpty()) {
            return this.s2.peek();
        }
        return this.front;
    }
    
    public boolean empty() {
        return this.s1.isEmpty() && this.s2.isEmpty();
    }

    public static void main(String[] args) {
        System.out.println("Hello, World!"); 
        MyQueue myQueue = new MyQueue();
        myQueue.push(1);
        myQueue.push(2);
        System.out.println(myQueue.peek());
        System.out.println(myQueue.pop());
        System.out.println(myQueue.empty());
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */